Fixed #27670 -- Prevented shell command from masking ImportErrors.

inglesp opened the pull request:
[Ticket](https://code.djangoproject.com/ticket/27670)

adamchainz commented on tests/shell/tests.py:
```diff
@@ -2,10 +2,24 @@
 
+try:
+    import IPython
+    del IPython
```
I don't see why these `del`s are needed - just a vague worry about leaving them in the module namespace?

timgraham replied:
The usual pattern is 
```
try:
    import IPython
except ImportError:
   IPython = None
```
No need for an extra variable.

adamchainz commented on django/core/exceptions.py:
```diff
@@ -192,3 +192,8 @@ def __repr__(self):
     pass
+
+
+class NoSuchShell(Exception):
```
I don't know if this needs to be a core exception, it seems too specific to the `shell` command to me

timgraham replied:
Agreed -- putting that in `shell.py` should be fine. Maybe that's better to inherit from `ImportError` and call it `ShellImportError`?

timgraham commented on django/core/exceptions.py:
```diff
@@ -192,3 +192,8 @@ def __repr__(self):
+
+
+class NoSuchShell(Exception):
+    """A shell does not exist."""
```
does not exist -> isn't installed?

timgraham commented on django/core/management/commands/shell.py:
```diff
@@ -127,11 +130,17 @@ def handle(self, **options):
-        raise ImportError("Couldn't load any of the specified interfaces.")
+            except NoSuchShell:
+                raise CommandError("Couldn't load {} interface.".format(shell))
+
```
chop blank line

timgraham commented on tests/shell/tests.py:
```diff
@@ -29,3 +43,13 @@ def test_stdin_read(self, select):
+
+    @unittest.skipIf(ipython_installed, 'IPython is installed')
+    def test_shell_with_ipython_not_installed(self):
+        with self.assertRaises(CommandError):
```
Please use `assertRaisesMessage()`.

timgraham commented on django/core/management/commands/shell.py:
```diff
@@ -127,11 +130,17 @@ def handle(self, **options):
+                    return getattr(self, shell)(options)
+                except NoSuchShell:
+                    pass
+            raise CommandError("Couldn't load any interface.")
```
I think it's impossible to reach this since the `python` shell doesn't raise `NoSuchShell`.

timgraham commented on django/core/management/commands/shell.py:
```diff
@@ -127,11 +130,17 @@ def handle(self, **options):
-                pass
-        raise ImportError("Couldn't load any of the specified interfaces.")
+            except NoSuchShell:
+                raise CommandError("Couldn't load {} interface.".format(shell))
```
Maybe "Couldn't import <shell>." would be slightly better wording.

knbk commented on django/core/management/commands/shell.py:
```diff
@@ -64,11 +65,13 @@ def ipython(self, options):
                 return
-        # no IPython, raise ImportError
-        raise ImportError("No IPython")
+        raise NoSuchShell("No IPython")
```
This still hides any `ImportError` if it's raised during the startup of IPython. I think you need to raise `NoSuchShell` when importing specific IPython versions fails, but before trying to initialize the shell, and let any other errors during initialization (including `ImportError`) bubble up.

inglesp commented:
Thanks @adamchainz, @timgraham, and @knbk for your reviews.  I think I've addressed all the comments.

However, there's a problem with the new tests, in that they pass when run individually or as part of the full test suite with `--parallel=1`, but as soon as tests run in parallel, the tests fail.  This is because something (I haven't worked out what yet) is closing stdin, causing the `select` [here](https://github.com/inglesp/django/blob/09823383190665f5cffa8dd0a9fc85f3a4c17b22/django/core/management/commands/shell.py#L142) to return something, which means we never get to the new code.

Is there a way for a test to detect whether it's being run in parallel?  If so, could we skip these tests in that case?

adamchainz commented on django/core/management/commands/shell.py:
```diff
@@ -39,36 +44,47 @@ def add_arguments(self, parser):
+        try:
+            from IPython.Shell import IPShell
+        except ImportError:
+            raise ShellImportError
```
I think it's preferable to raise instances rather than classes. Python 3 will instantiate it for you when required anyway ([docs](https://docs.python.org/3/reference/simple_stmts.html#the-raise-statement)):

> Otherwise, raise evaluates the first expression as the exception object. It must be either a subclass or an instance of BaseException. If it is a class, the exception instance will be obtained when needed by instantiating the class with no arguments.

timgraham replied:
Adam, I guess you meant "classes rather than instances" since this comment isn't addressed?

adamchainz replied:
I meant I thought it should change to `raise ShellImportError()` to be clearer

adamchainz commented:
> If so, could we skip these tests in that case?

I think we should just fix the fundamental problem. It was probably only added 14 days ago in #7681 , maybe not detected then because there weren't any other tests running `shell` after ones it added.

inglesp commented:
Would changing

```
 if sys.platform != "win32" and select.select([sys.stdin], [], [], 0)[0]:
    exec(sys.stdin.read())
    return
```

to

```
if sys.platform != "win32" and select.select([sys.stdin], [], [], 0)[0]:
    data = sys.stdin.read()
    if data:
        exec(data)
        return
```

be an adequate fix?

Or should we aim to fix the problem of stdin getting closed when tests are run in parallel?

adamchainz commented:
I think your change is against the spirit of #7681 , which was trying to bring `./manage.py shell` in line with `python`, which still "executes" an empty `stdin` and exits:

```sh
$ python <<< ''
$
```

That said it's such an edge case it might be more convenient to just do it. Fixing the parallel test environment to match the non-parallel one is likely to prevent future similar bugs with `stdin` of course.

inglesp commented:
The root cause of the problem is that running tests in parallel uses the multiprocessing module which [closes the sys.stdin for subprocesses and replaces it with os.devnull](https://github.com/python/cpython/blob/master/Lib/multiprocessing/util.py#L364-L381).  Since [reading from /dev/null always returns EOF](http://man7.org/linux/man-pages/man4/null.4.html), when we run our tests in parallel,`select` will always show that `sys.stdin` is readable.

There's no way to detect whether you're reading from /dev/null or not, so I don't think this problem is fixable in the code.

Instead, I've patched `select.select()` in the tests to indicate that there's nothing to read on `sys.stdin`.  I think this is the best we can do.

adamchainz commented on tests/shell/tests.py:
```diff
@@ -29,3 +39,17 @@ def test_stdin_read(self, select):
         self.assertEqual(stdout.getvalue().strip(), '100')
+
+    @unittest.skipIf(IPython is not None, 'IPython is installed')
+    @mock.patch('django.core.management.commands.shell.select.select')
```
Leave a comment about your finding with parallel tests so future test authors for `shell` don't have to re-learn it all. Perhaps even extract the patching into `setUp()` / `tearDown()` of the whole test class so it's automatic

inglesp replied:
Have added a comment.  Putting the patching in `setUp()` won't work since we don't want to patch for `test_stdin_read`.

adamchainz replied:
Could unpatch in just that one test

inglesp commented:
> Is another workaround in the line that does the `select.select` possible, checking if `sys.stdin` is pointed at `os.devnull`? Perhaps with `sys.stdin.fileno()` or similar?

This wouldn't work without a lot of hassle -- you can't easily go from a file descriptor to the name of a file.  (And `os.devnull` is just an alias for the filename `'/dev/null'` on POSIX and `'nul'` on Windows.)

You could test for `sys.stdin.fileno()` being zero, in which case it's definitely stdin, but this would cause problems where eg a user runs the command via the `subprocess` module and passes in their own `stdin`.

adamchainz commented:
Makes sense. Test failures all seem to be widget rendering ones

timgraham commented:
I'm not sure if I like the approach in general. Would it be possible to make a change in the `python` method to raise something other than `ImportError` for the problematic case you discovered so that same boilerplate isn't needed in the other methods?

adamchainz commented:
@timgraham It's not possible as far as I understand because the import errors we're trying to avoid masking occur in the non import lines e.g. `app = TerminalIPythonApp.instance()` and it's not possible in python to detect what exactly failed to import where, without doing horrible traceback inspection.

timgraham commented:
Okay, Peter perhaps you can confirm that's the case -- it's not clear to me from the ticket exactly where the problem is. Anyway, that does give me the idea to first remove support for IPython < 1.0 to simplify this a bit: https://github.com/django/django/pull/7821.

inglesp commented:
The original `ImportError` was actually raised when using the `python` interface [when `exec()`ing my `.pythonrc.py` file](https://github.com/django/django/blob/master/django/core/management/commands/shell.py#L80).

I should have thought through this properly sooner, but I think the fix is actually to catch any exception raised by that `exec`, print the traceback to screen, and carry on.  This would match the behaviour of the normal Python prompt, as well as that of bpython and IPython.

I'll close this PR and open two more: one to fix this problem, and one to provide a better error message when you try to use a shell that's not installed.
